File: Minimum_Pair_Removal_to_Sort_Array_I/minimum_pair_removal_to_sort_array_i.py

# COMPLETE UN-OPTIMIZED 1st Attempt Brute Force


class Solution:


    def minimumPairRemoval(self, nums: List[int]) -> int:
        if len(nums) == 1:
            return 0
        
        #Example [5,2,3,1]
        # [5,2,4]
        # [5,6]
        #
        # [5,1,7] -> [6,7]


        # [2,-100,105]
        # [-98, 105]


        # [1,2,3,-100]
        # Steps: check if it Decreases at all
        # 
        # adjacent = []
        moves = 0
        while (self.check_order(nums) == False): # and moves < 5 ):
            # calculate adjacency list

            min_val = float('inf')
            for i in range(1, len(nums)):
                value = nums[i]+nums[i-1]

                min_val = value if value < min_val else min_val

            moves += 1

            # search for indices in nums, and combine them

            curr = 1

            while (nums[curr]+nums[curr-1]!=min_val):
                curr += 1

            nums[curr-1] = min_val
            nums.pop(curr)
        return moves



    def check_order(self, nums) -> bool:

        curr = 1
        prev = 0
        while (curr < len(nums)):
            if nums[curr] < nums[prev]:
                return False
            curr += 1
            prev += 1
        return True

File: Minimum_Pair_Removal_to_Sort_Array_I/minimum_pair_removal_to_sort_array_i.cpp
#include <stdio.h>
#include <vector>

using namespace std;

class Solution {
    public:
        int minimumPairRemoval(vector<int>& nums) {
    
            int moves = 0;
    
            while (check_order(nums) == false) {

                int min_val = INT_MAX;
                int curr = 1;
    
                for (int i = 1; i < nums.size(); i++ ) {
                    int value = nums[i] + nums[i-1];
                    curr = value < min_val ? i : curr;
                    min_val = value < min_val ? value : min_val;
                }
                moves++;
    
                nums[curr-1] = min_val;
                nums.erase(nums.begin() + curr);
            }
    
    
    
            return moves;
            
        }
    
        bool check_order(vector<int>& nums) {
    
            for (int i = 1; i < nums.size(); i++) {
                if (nums[i] < nums[i - 1]) {
                    return false;
                }
            }
    
            return true;
        }
    };

int main() {
    vector<int> tv1 = {1,2,3,4,5};
    vector<int> tv2 = {5,4,3,2,1};
    vector<int> tv3 = {1,1,4,2};

    int a1, a2, a3;

    Solution s1;

    a1 = s1.minimumPairRemoval(tv1);
    a2 = s1.minimumPairRemoval(tv2);
    a3 = s1.minimumPairRemoval(tv3);

    printf("a1 = %d\n", a1);
    printf("a2 = %d\n", a2);
    printf("a3 = %d\n", a3);


    return 0;


}